File: haskell/game.hs
module ConnectK where

import Data.List (transpose, tails, isPrefixOf)
import Data.Maybe (isJust)

type Player = Int
type Board = [[Int]]
data Params = Params { nP :: Int, mP :: Int, kP :: Int }

-- Ініціалізація та хід
makeMove :: Board -> Int -> Player -> Board
makeMove board col player =
    let row = getDropRow board col
    in case row of
        Nothing -> board
        Just r  -> replace2D board r col player

getDropRow :: Board -> Int -> Maybe Int
getDropRow board col =
    let column = map (!! col) board
        emptyRows = [r | (r, val) <- zip [0..] column, val == 0]
    in if null emptyRows then Nothing else Just (last emptyRows)

replace2D :: Board -> Int -> Int -> Int -> Board
replace2D b r c val =
    let (before, targetRow:after) = splitAt r b
        newRow = take c targetRow ++ [val] ++ drop (c + 1) targetRow
    in before ++ [newRow] ++ after

-- Перевірка перемоги
isWin :: Params -> Board -> Player -> Bool
isWin p b player = any (hasK (kP p) player) (allLines b)
  where
    hasK k ply line = any (replicate k ply `isPrefixOf`) (tails line)
    allLines bd = bd ++ transpose bd ++ diags bd ++ diags (map reverse bd)
    diags bd = transpose [replicate i 0 ++ r | (i, r) <- zip [0..] bd]

-- Евристична оцінка
evaluate :: Params -> Board -> Int
evaluate p b = sum $ map scoreWindow (allWindows (kP p) b)
  where
    allWindows k bd = concatMap (getWindows k) (allLines bd)
    
    getWindows k line
        | length line < k = []
        | otherwise = take k line : getWindows k (tail line)
        
    allLines bd = bd ++ transpose bd ++ diags bd ++ diags (map reverse bd)
    diags bd = transpose [replicate i (-1) ++ r | (i, r) <- zip [0..] bd] 
    
    scoreWindow win
        | (-1) `elem` win = 0 
        | 1 `elem` win && 2 `elem` win = 0 
        | otherwise = (length (filter (==1) win))^2 - (length (filter (==2) win))^2

-- Alpha-Beta MinMax
alphaBeta :: Params -> Board -> Int -> Int -> Int -> Player -> (Int, Maybe Int)
alphaBeta p b depth alpha beta player
    | depth == 0 = (evaluate p b, Nothing)
    | null moves = (0, Nothing)
    | player == 1 = maximize moves alpha beta (-2000000) Nothing
    | otherwise   = minimize moves alpha beta 2000000 Nothing
  where
    moves = [c | c <- [0..(mP p - 1)], isJust (getDropRow b c)]

    maximize [] _ _ bestV bestM = (bestV, bestM)
    maximize (m:ms) a b bestV bestM =
        let nextB = makeMove b m 1
            score = if isWin p nextB 1 then 1000000 + depth 
                    else fst $ alphaBeta p nextB (depth - 1) a b 2
            (nV, nM) = if score > bestV then (score, Just m) else (bestV, bestM)
            nA = max a nV
        in if b <= nA then (nV, nM) else maximize ms nA b nV nM

    minimize [] _ _ bestV bestM = (bestV, bestM)
    minimize (m:ms) a b bestV bestM =
        let nextB = makeMove b m 2
            score = if isWin p nextB 2 then -1000000 - depth 
                    else fst $ alphaBeta p nextB (depth - 1) a b 1
            (nV, nM) = if score < bestV then (score, Just m) else (bestV, bestM)
            nB = min b nV
        in if nB <= a then (nV, nM) else minimize ms a nB nV nM

File: python/game.py
import copy

class ConnectK:
    def __init__(self, n, m, k, forbidden=None):
        self.n = n
        self.m = m
        self.k = k
        # 0: порожньо, 1: гравець 1, 2: гравець 2, -1: заборонено
        self.board = [[0 for _ in range(m)] for _ in range(n)]
        if forbidden:
            for r, c in forbidden:
                if 0 <= r < n and 0 <= c < m:
                    self.board[r][c] = -1

    def get_valid_moves(self):
        """Повертає список стовпців, де верхня клітинка вільна."""
        return [c for c in range(self.m) if self.board[0][c] == 0]

    def get_drop_row(self, col):
        """Ефект падіння: знаходимо найнижчий вільний рядок."""
        for r in range(self.n - 1, -1, -1):
            if self.board[r][col] == 0:
                return r
        return None

    def make_move(self, col, player):
        row = self.get_drop_row(col)
        if row is not None:
            self.board[row][col] = player
            return row
        return None

    def check_win(self, r, c, p):
        """Повна перевірка перемоги у 4-х напрямках."""
        directions = [(0, 1), (1, 0), (1, 1), (1, -1)]
        for dr, dc in directions:
            count = 1
            for direction in [1, -1]:
                nr, nc = r + dr * direction, c + dc * direction
                while 0 <= nr < self.n and 0 <= nc < self.m and self.board[nr][nc] == p:
                    count += 1
                    nr += dr * direction
                    nc += dc * direction
            if count >= self.k: return True
        return False

    def evaluate(self):
        """Просунута евристика: оцінка всіх ліній довжиною K."""
        score = 0
        for r in range(self.n):
            for c in range(self.m):
                score += self._count_window_score(r, c)
        return score

    def _count_window_score(self, r, c):
        total = 0
        for dr, dc in [(0, 1), (1, 0), (1, 1), (1, -1)]:
            window = []
            for i in range(self.k):
                nr, nc = r + dr * i, c + dc * i
                if 0 <= nr < self.n and 0 <= nc < self.m:
                    window.append(self.board[nr][nc])
                else:
                    break
            if len(window) == self.k:
                total += self._calculate_heuristic(window)
        return total

    def _calculate_heuristic(self, window):
        p1 = window.count(1)
        p2 = window.count(2)
        if p1 > 0 and p2 > 0: return 0 
        if p1 > 0: return 10 ** (p1 - 1)
        if p2 > 0: return -(10 ** (p2 - 1))
        return 0

def alpha_beta(game, depth, alpha, beta, player):
    valid_moves = game.get_valid_moves()
    if depth == 0 or not valid_moves:
        return game.evaluate(), None

    best_move = valid_moves[0]
    if player == 1: 
        max_eval = -float('inf')
        for move in valid_moves:
            temp_game = copy.deepcopy(game)
            r = temp_game.make_move(move, 1)
            if temp_game.check_win(r, move, 1):
                eval = 1000000 + depth
            else:
                eval, _ = alpha_beta(temp_game, depth - 1, alpha, beta, 2)
            if eval > max_eval:
                max_eval, best_move = eval, move
            alpha = max(alpha, eval)
            if beta <= alpha: break
        return max_eval, best_move
    else: 
        min_eval = float('inf')
        for move in valid_moves:
            temp_game = copy.deepcopy(game)
            r = temp_game.make_move(move, 2)
            if temp_game.check_win(r, move, 2):
                eval = -1000000 - depth
            else:
                eval, _ = alpha_beta(temp_game, depth - 1, alpha, beta, 1)
            if eval < min_eval:
                min_eval, best_move = eval, move
            beta = min(beta, eval)
            if beta <= alpha: break
        return min_eval, best_move